src/library/scala/collection/immutable/ArraySeq.scala

```diff
@@ -208,18 +208,9 @@ object ArraySeq extends StrictOptimizedClassTagSeqFactory[ArraySeq] { self =>
 
   def empty[A : ClassTag]: ArraySeq[A] = emptyImpl
 
-  def from[A : ClassTag](it: scala.collection.IterableOnce[A]): ArraySeq[A] = unsafeWrapArray {
-    val n = it.knownSize
-    if (n > -1) {
-      val elements = Array.ofDim[A](n)
-      val iterator = it.iterator
-      var i = 0
-      while (i < n) {
-        ScalaRunTime.array_update(elements, i, iterator.next())
-        i = i + 1
-      }
-      elements
-    } else ArrayBuffer.from(it).toArray
+  def from[A](it: scala.collection.IterableOnce[A])(implicit tag: ClassTag[A]): ArraySeq[A] = it match {
+    case as: ArraySeq[A] if as.elemTag.runtimeClass isAssignableFrom tag.runtimeClass => as
```

Comment:
I don't quite understand why we need the `isAssignableFrom` condition.  Because `ArraySeq` is covariant, it has to tolerate widening of its element type.  So if `tag` is too wide (not assignable), `ArraySeq` has to cope anyway because you could instead do the `from` on the narrow type and then upcast.

That is, these two things should be equivalent:

```
ArraySeq.from(ArraySeq[X]: Seq[Y])
ArraySeq.from(ArraySeq[X]): ArraySeq[Y]
```

So I don't think it would be any worse to just drop this condition?

(Everything else LGTM.)

Reply:
that is a very good point. I'll remove it

Reply:
my only concern would be if you explicitly typed it as a primitive and expected it to return an `ArraySeq` backed by a primitive `Array`, but instead it returned one backed by a `Array[T <: AnyRef]`.

Reply:
If it's a primitive, you shouldn't have `T <: AnyRef` to begin with unless you've cast it, and then it's the casting that's unsound, not this.  (FWIW, the flexibility that `ArraySeq` displays with its underlying array type vs. its stated element type worries me, but I have yet to find an unsoundness.  I haven't looked very hard, though.)

Reply:
but you can have an `ArraySeq[Int]` backed by an `Array[AnyRef]` (technically `Array[Any]`, but same thing).

It doesn't help that all of the operations return an `ArraySeq` backed by an `Array[Any]` anyway.

Reply:
All IterableCC-based operations have to box primitives but others should be able to build primitive arrays where you expect them.

Reply:
TODO: open a ticket to optimise implementations for operations on the same type/a subtype (e.g. updated)